### PR TITLE
Feature 94 rotation Jacobians

### PR DESCRIPTION
**Pull request by arunxdas**
This PR adds the following functions to the geometry library for rotations:

1) logMapAndJacobian
2) inverseAndJacobian
3) composeAndJacobian
4) manifoldMinusAndJacobian
5) overloaded -operator

It also modifies the numerical differentiation to perform forward difference on manifold quantities

### CODE REVIEW

File: wave_geometry/include/wave/geometry/numerical_test_functors.hpp [hunk 1]
```diff
@@ -40,13 +40,104 @@ class RotateAndJacobianJparamFunctor {
...
+        this->R_left = R_left;
+        this->R_right = R_right;
+    }
 
-// Implement the numerical differentiation, following procedure
-// for a 5 point stencil https://en.wikipedia.org/wiki/Five-point_stencil
-// Exact step size representation is outlined in
-// "Numerical Recipes by Brian P. Flannery,
-// Saul Teukolsky, and William H. Press, chapter 5.7.
+    Rotation operator()(Vec3 perturbation) {
```
Comment (chutsu) [inline wave_geometry/include/wave/geometry/numerical_test_functors.hpp]: `const Vec3 &` ?

File: wave_geometry/include/wave/geometry/numerical_test_functors.hpp [hunk 2]
```diff
@@ -40,13 +40,104 @@ class RotateAndJacobianJparamFunctor {
...
+        R_perturbed = R_perturbed.manifoldPlus(perturbation);
+        return R_perturbed.composeAndJacobian(R_right, J, J);
+    }
+};
 
+class ComposeAndJacobianJRightFunctor {
+ public:
+    Rotation R_left;
+    Rotation R_right;
+    ComposeAndJacobianJRightFunctor(Rotation R_left, Rotation R_right) {
```
Comment (chutsu) [inline wave_geometry/include/wave/geometry/numerical_test_functors.hpp]: `const Rotation &R_left`?

File: wave_geometry/include/wave/geometry/numerical_test_functors.hpp [hunk 3]
```diff
@@ -40,13 +40,104 @@ class RotateAndJacobianJparamFunctor {
...
+ public:
+    Rotation R_left;
+    Rotation R_right;
+    ManifoldMinusAndJacobianJLeftFunctor(Rotation R_left, Rotation R_right) {
+        this->R_left = R_left;
+        this->R_right = R_right;
+    }
+
+    Vec3 operator()(Vec3 perturbation) {
+        Mat3 J;
```
Comment (chutsu) [inline wave_geometry/include/wave/geometry/numerical_test_functors.hpp]: Where is `J` used?

File: wave_geometry/include/wave/geometry/numerical_test_functors.hpp [hunk 4]
```diff
@@ -58,42 +149,36 @@ double get_perturbation_point(double evaluation_point, double step_size) {
     return evaluation_point + exact_step_size;
 }
 
+
+// Numerical difference using simple forward difference method.
+// Requires overloaded -operator to perform manifoldMinus for
+// Manifold quantities.
```
Comment (chutsu) [inline wave_geometry/include/wave/geometry/numerical_test_functors.hpp]: This is not quite doxygen style, check out the measurement container

File: wave_geometry/src/rotation.cpp [hunk 5]
```diff
@@ -116,13 +116,81 @@ Vec3 Rotation::logMap() const {
...
+                                      Mat3 &J_left,
+                                      Mat3 &J_right) const {
+    checkMatrixFinite(R_right.toRotationMatrix());
+
+    // Compute the rotation composition.
+    Rotation R_out;
+    R_out = (*this) * R_right;
+
+    // Jacobian wrt left rotation is identity.
+    J_left = Eigen::MatrixXd::Identity(3, 3);
```
Comment (chutsu) [inline wave_geometry/src/rotation.cpp]: `MatX::Identity()` also works

File: wave_geometry/src/rotation.cpp [hunk 6]
```diff
@@ -116,13 +116,81 @@ Vec3 Rotation::logMap() const {
...
+
+    // Jacobian wrt right rotation is the left rotation matrix.
+    J_right = this->toRotationMatrix();
+    ;
```
Comment (chutsu) [inline wave_geometry/src/rotation.cpp]: opps

**Comment by chutsu** [comment]
Update your docs to doxygen style comments

File: wave_geometry/include/wave/geometry/numerical_test_functors.hpp [hunk 7]
```diff
@@ -40,13 +40,104 @@ class RotateAndJacobianJparamFunctor {
     }
 };
 
+class ComposeAndJacobianJLeftFunctor {
+ public:
+    Rotation R_left;
+    Rotation R_right;
+    ComposeAndJacobianJLeftFunctor(Rotation R_left, Rotation R_right) {
```
Comment (cglwn) [inline wave_geometry/include/wave/geometry/numerical_test_functors.hpp]: Make inputs const ref?

ComposeAndJacobianJLeftFunctor(const Rotation& R_left, const Rotation& R_right) {

File: wave_geometry/include/wave/geometry/numerical_test_functors.hpp (hunk 1 above)
Comment (arunxdas) [inline wave_geometry/include/wave/geometry/numerical_test_functors.hpp]: done

File: wave_geometry/include/wave/geometry/numerical_test_functors.hpp (hunk 2 above)
Comment (arunxdas) [inline wave_geometry/include/wave/geometry/numerical_test_functors.hpp]: done for all inputs

File: wave_geometry/include/wave/geometry/numerical_test_functors.hpp (hunk 3 above)
Comment (arunxdas) [inline wave_geometry/include/wave/geometry/numerical_test_functors.hpp]: it's a "dummy" variable that is passed into composeAndJacobian(R_right, J, J);

We don't use J, only the return value of composeAndJacobian.  In a future PR, we should check for null arguments on the jacobians and only compute them if asked for

File: wave_geometry/src/rotation.cpp (hunk 5 above)
Comment (arunxdas) [inline wave_geometry/src/rotation.cpp]: done

File: wave_geometry/src/rotation.cpp (hunk 6 above)
Comment (arunxdas) [inline wave_geometry/src/rotation.cpp]: haha, fixed

File: wave_geometry/include/wave/geometry/numerical_test_functors.hpp (hunk 7 above)
Comment (arunxdas) [inline wave_geometry/include/wave/geometry/numerical_test_functors.hpp]: yup, done for all inputs in file

File: wave_geometry/include/wave/geometry/numerical_test_functors.hpp (hunk 4 above)
Comment (arunxdas) [inline wave_geometry/include/wave/geometry/numerical_test_functors.hpp]: Done, comments are doxygen style now

File: wave_geometry/include/wave/geometry/numerical_test_functors.hpp [hunk 8]
```diff
@@ -33,26 +33,122 @@ class RotateAndJacobianJparamFunctor {
...
+
+class ManifoldMinusAndJacobianJLeftFunctor {
+ public:
+    Rotation R_left;
+    Rotation R_right;
+    ManifoldMinusAndJacobianJLeftFunctor(const Rotation &R_left,
+                                         const Rotation &R_right) {
+        this->R_left = R_left;
+        this->R_right = R_right;
+    }
```
Comment (chutsu) [inline wave_geometry/include/wave/geometry/numerical_test_functors.hpp]: No big deal but in future prefer defining constructors via list initialization: https://stackoverflow.com/questions/926752/why-should-i-prefer-to-use-member-initialization-list

**Pull request merged**